core/rapid_test.go

```diff
@@ -0,0 +1,162 @@
+				return nodes[nodeIndex]
+			}
+
+			// Make sure the only proposer is picked using Round Robin
+			backend.isProposerFn = func(from []byte, height uint64, _ uint64) bool {
+				return bytes.Equal(from, nodes[height%numNodes])
+			}
+
+			// Make sure the proposal is valid if it matches what node 0 proposed
+			backend.isValidBlockFn = func(newProposal []byte) bool {
+				return bytes.Equal(newProposal, proposal)
+			}
+
+			// Make sure the proposal hash matches
+			backend.isValidProposalHashFn = func(p []byte, ph []byte) bool {
+				return bytes.Equal(p, proposal) && bytes.Equal(ph, proposalHash)
+			}
+
+			// Make sure the preprepare message is built correctly
+			backend.buildPrePrepareMessageFn = func(
+				proposal []byte,
+				certificate *proto.RoundChangeCertificate,
+				view *proto.View,
+			) *proto.Message {
+				return buildBasicPreprepareMessage(
+					proposal,
+					proposalHash,
+					certificate,
+					nodes[nodeIndex],
+					view)
+			}
+
+			// Make sure the prepare message is built correctly
+			backend.buildPrepareMessageFn = func(proposal []byte, view *proto.View) *proto.Message {
+				return buildBasicPrepareMessage(proposalHash, nodes[nodeIndex], view)
+			}
+
+			// Make sure the commit message is built correctly
+			backend.buildCommitMessageFn = func(proposal []byte, view *proto.View) *proto.Message {
+				return buildBasicCommitMessage(proposalHash, committedSeal, nodes[nodeIndex], view)
+			}
+
+			// Make sure the round change message is built correctly
+			backend.buildRoundChangeMessageFn = func(
+				proposal []byte,
+				certificate *proto.PreparedCertificate,
+				view *proto.View,
+			) *proto.Message {
+				return buildBasicRoundChangeMessage(proposal, certificate, view, nodes[nodeIndex])
+			}
+
+			// Make sure the inserted proposal is noted
+			backend.insertBlockFn = func(proposal []byte, _ []*messages.CommittedSeal) {
+				insertedBlocks[nodeIndex][currentProposals[nodeIndex]] = proposal
+				currentProposals[nodeIndex]++
+			}
+
+			// Make sure the proposal can be built
+			backend.buildProposalFn = func(u uint64) []byte {
+				return proposal
+			}
+		}
+
+		// Initialize the backend and transport callbacks for
+		// each node in the arbitrary cluster
+		backendCallbackMap := make(map[int]backendConfigCallback)
+		transportCallbackMap := make(map[int]transportConfigCallback)
+
+		for i := 0; i < int(numNodes); i++ {
+			i := i
+			backendCallbackMap[i] = func(backend *mockBackend) {
+				commonBackendCallback(backend, i)
+			}
+
+			transportCallbackMap[i] = commonTransportCallback
+		}
+
+		// Create the mock cluster
+		cluster := newMockCluster(
+			numNodes,
+			backendCallbackMap,
+			nil,
+			transportCallbackMap,
+		)
+
+		// Set the multicast callback to relay the message
+		// to the entire cluster
+		multicastFn = func(message *proto.Message) {
+			cluster.pushMessage(message)
+		}
+
+		// Run the sequence up until a certain height
+		cluster.runSequenceUntilHeight(desiredHeight)
```

Comment:
I think it may be enough just to create a block. We need to check that all honest nodes will be in the same state in the end.

Reply:
You're right - part of the test definitely relies on the proposal passing all validations in runtime (otherwise the sequence would not finish gracefully). We make sure at the end that the inserted proposals are the same for each node in the cluster:
https://github.com/0xPolygon/go-ibft/blob/2ae44e79463befc2cfb25781feff11453aa55875/core/rapid_test.go#L190-L198

Reply:
@b00ris  ping to resolve this convo if the discussion was answered